### ISSUE TITLE
Garment Color & Change Product functionality

### ISSUE THREAD

**Issue by NullOperator404** [NullOperator404/Developer-Diary#27]
UPDATE 1/25/23:

The garment color functionality is now in place. Presently, a JavaScript object exists within the code that provides data for color swatches and functions. The object has 4 key-value pairs: 
**name:** - This is a String with proper capitalization. JS uses this value for displaying the color name in the DOM.
**color:** - This value is the accompanying hex value for the color. JS uses this to fill the swatch. 
**front:** - This value is a filename that refers to the front image of the garment in the corresponding color. JS uses this value to update the image in the design space.
**back:** - This value is the same as the `front`, only it's for the back view of the garment. JS updates both views simultaneously.

However, it was obvious from the start that the entire dataset could not exist in the JS code stack, as this would balloon the file size and thus slow down responsiveness for users. The problem here is identical to the problem I faced a few weeks ago with the clipart; the client-side needs to only download what it *needs,* and so the solution seems to be simple: a JSON file will need to be setup containing objects representing products. JS will then `Fetch` the necessary data from this file. 

With this revelation has come another problem, though. The issue now is how to best organize all of the product data into an efficient database that JS can access/read. When working on the color select functionality, I originally thought it might be best to have individual JSON files for each product, but this now seems inefficient. 

A more efficient approach would be to build a single JSON file containing a single `array` of objects. Each object will represent a particular product, and will contain key-value pairs for any and all data the browser might need at some point for display purposes. For example, a specific object might contain the product brand, product number, style, size, material, color names, color codes, descriptive text to be used in the DOM, and more. 

While this approach sounds good in theory, putting it into practice is not so straightforward. Writing JSON is extremely time-consuming, and writing one in Notepad is not nearly as controllable as using an actual database. However, JSON would be the easiest to work with within JS, as it would avoid any language issues that are often the case with cross-language interfaces like JS-to-PHP or JS-to-MySQL. 

The main issue is that there are hundreds of products, and so this database must be built before I can move forward on the development of the lab. At present, I am working on the "Change Product" feature, which will rely heavily on information retrieved from the database. 

I will look further into this issue before committing to an approach. GoDaddy - where the site is hosted - offers databases for the website, but working with their online-interface is a bit clunky; I'd rather build the database locally so it could be tested first. Then move it to the server for remote testing.

**Comment by NullOperator404** [comment]
ADDENDUM:
The biggest issue I'm facing at the moment (apart from the sheer scale of the database) is how to best implement access to the data without bogging down user experience. JSON isn't going to work, and it's now looking more and more like I may be relying on PHP again. Ugh. 

The reason is because if I setup a JSON file containing all of the products, then the client-side will then have to parse the entire database before it can select data sets from it. While I don't anticipate the file size to be over 500 kb, there really is no need for client-side to handle the entire database operation. The database needs to remain server-side at all times, and the client just retrieve a particular object (G500, for example) which would then populate an empty var with this object data. Now the client has immediate access to all data pertaining to the G500 product. Since the G500 object data set would then be loaded client-side, any further operations could access the data directly without needing to contact the server. The server would only need to be contacted if the PRODUCT changes. If the user changes to say, G640, the browser would then need to contact the server to retrieve the data set for G640. The object var would then be overwritten with the new dataset. 

The alternative of course, is to set up separate JSON files for each individual product. Would this be more efficient? 

In this approach, the client would call a specific JSON file based on product selected. The JSON would populate a local array, which would then be used for all operations (until the user changes products, at which point the array would need to be updated). 

Hmm...selective JSON files might actually be *faster,* since neither the client nor server would ever have to sift through an entire database. Instead, JS would call a particular JSON file based on product selected. This would mean that AT MOST, the client would only need to retrieve 3 to 4 KB. This would be virtually instantaneous on any connection speed or device. Plus, once the JSON file has been called, it would then be locally accessible, eliminating any further delays. 

The issue here is that it's going to be more work on my end, but the payoffs are obvious. Easy to work with, easy to organize, and is probably the quickest way to access/process data.

**Comment by NullOperator404** [comment]
UPDATE 1/26/23, 3:14 am

After a considerable amount of time twisting my brain into a pretzel, I've worked out a system that I think will be the most efficient for both client & server. 

For this system to work, each product will have it's own JSON file as I suggested. However, there was also the issue of building these JSON files, and having to input numerous key-value pairs that would be the same across numerous files. To resolve this issue, I have devised a method that will not only reduce the size of each JSON file, but will also save me having to retype the same key-value pairs over and over for different products. So the system works like this:

Within the main JS file exists several arrays containing values that never change. For example, an array called `brands` contains string values for each brand offered - Gildan, Comfort Colors, Next Level, etc. 

In the product JSON file, which contains a data object, the value for the "brand" key will be a NUMBER. The number will correspond to an index in the array. So if "Gildan" is at index 0, then any products with the brand "Gildan" will have "0" as the value for their "brand" keys. 

Back in JS, whenever the brand name needs to be retrieved and used, JS will use the number values in the JSON file to reference the index in the array, thus retrieving the brand name. 

This system will be most effective when dealing with colors, as each brand offers a huge variety of colors and it would take weeks, if not months, to type these name/hex values over and over for each product. Instead, I've set up an array `masterColors` which contains objects. Each object has two key-value pairs: a `name` and a `hex` value. The `name` key refers to a string which JS will use for visible text displays in the DOM, such as the color palette panel. The `hex` key refers to corresponding hex values which will be used to apply fill colors to the swatches. 

Perhaps best of all though is that I can now be selective with colors from `masterColors` by specifying a series of index values in a sub-array. For instance, 5-6-7-10-12-15 could be used to select specific colors from the master array. This will come in handy since some styles are only offered in certain colors. 

Eventually, I will likely transfer all of this array code to it's own file so as not to clog up the main JS file. This JS object array will likely end up containing hundreds of objects, as I will need to include brand-specific colors for all brands offered. 

This does leave one other hurdle that I will need to address later: filenames. Unfortunately, each color for each garment will need to have it's own image pair (front/back), meaning that the total number of image files will be TWICE as many colors listed in the array. I'm not 100% sure yet how I can synchronize filename specification with the dynamically-generated palette picker. For now, I'm going to focus on the `masterColor` array and getting it fully populated.

**Comment by NullOperator404** [comment]
UPDATE 1/27/23

I've built a large color array and moved the code to a separate JS file. I've also started working on building out the JSON file that uses the reference numbers to the different arrays. This method seems to work well, particularly with the color swatches. It's super easy to keep the two respective color key-values synchronized since they should be identical; index 4 for example, contains a `name` and the `hex` value for the specified color. 

However, it just occurred to me moments earlier that one of my vendors may have a complete database already prepared and ready to download for business purposes, and it turns out they do. I am currently downloading the database to see if it can be utilized with my system. If not, it will at least save a ton of time retrieving garment color images, as they are all included in the database. 

The issue now is size. It's 400+ MB in and still isn't finished downloading. This could take a while...

Also, I feel almost certain that this database would be incompatible with the retrieval system I've setup, but we will see.

**Comment by NullOperator404** [comment]
The test database from my vendor is a spreadsheet format, so this isn't really much use to me. Also, the spreadsheet contains numerous data columns that are not needed for my project, so even if it were possible to export the file to a JSON format, it would contain a lot of unnecessary data that would only bog things down. 

For these reasons, I have elected to not use their DB and continue building my own in the JSON file, structuring it to work with the interface I've designed in Javascript. 

One thing that I will be using the vendor site's resource center for is acquiring all of the front/back images needed for the various product colors. 

I've got 27 items entered into the JSON file thusfar, so I think I'm going to break from data entries and start working on setting up the `Fetch` method that will call the JSON file. All of the logic needs to be setup that will utilize the various key-value pairs in the file. I also need to be sure the index-to-hex value system I've setup for color palettes will work correctly.

**Comment by NullOperator404** [comment]
At lot has happened in the past week, so I'll try to bring the diary up to speed. 

**IDE PROBLEM:** - Last weekend, I began preparing tests for the JSON database file and discovered that the `Fetch` request for the JSON file was throwing the same CORS error that the clipart function was. After a few days of troubleshooting, I realized that the IDE I'm using to code the project (IntelliJ) had stopped opening my HTML file on it's built-in server, localhost: 63342. Most of last week was spent playing email tag with tech support at JetBrains, trying to resolve this issue. Day before yesterday, the problem was solved by creating a new project folder and copying all of my project files into that new folder. Since I was unable to work on the JS itself, I spent my downtime working on building out the JSON product database file, as well as building the new color select system. 

The IDE problem also explains why the clipart was no longer loading (as it also uses a `Fetch` request). 

**JSON TESTS** - After fixing the localhost problem, I was once again able to test the `Fetch` request, and after setting up some HTTML/CSS to describe how the search results would look, the test was a complete success. Clicking the "Change Product" button now opens a pop-up window with cards for the 8 main categories - T-shirts, Hoodies, Sweatshirts, Dri-Fit, Jackets, Polos, Signs, and Decals. 

JS logic has been setup where the browser will only display items from the JSON file matching my specified criteria. For example, clicking "T-shirts" will only return any item from the database where the "Category" key has a value of "T-shirt." Later, I want to implement a search refinement system that will allow users to weed out unwanted brands, materials, sizes, etc. 

**COLOR CHANGE SYSTEM** - yesterday while finishing up some of the JSON tests, it occurred to me that my current "garment color change" system will simply be too labor-intensive and inefficient. In a nutshell, the current system would require 2 images for every color of every product. The Gildan G500 alone is offered in over 70 colors, so this would be 140 images just for this one t-shirt. Considering the complete product line will likely consist of 300+ items, this system just isn't feasible. It would mean the collection and editing of thousands of images. Furthermore, the browser would need to retrieve new image pairs from the server each time the user selects a new garment color. This too, would bog down the user experience, especially on slower connections. 

For these reasons, I decided to revisit my original proposal for the color-change system. I tested this method a few weeks ago before building the 2-image pair system, but abandoned it due to unforeseen side effects that I will explain below. 

This alternative system would utilize the CSS `filter` property to make adjustments to a pre-loaded image via `brightness()`, `hue-rotate()`, `contrast()`, `saturate()`, and `opacity.` The base image used would be a red garment to provide a relatable starting point on the color wheel. The `hue-rotate` function moves the color hue a specified number of degrees on a theoretical color wheel. Thus, other colors like greens, blues, yellows, browns, etc can be created from the default "red" image simply by adjusting the hue. The issue I ran into with my test was that since the `filter` property was being applied to the parent element of the design space, then every child element was also being affected. Thus, adjusting the hue would also affect the colors of any user-generated text or art elements in the design space. So at the time, I abandoned it. 

However, yesterday I realized that an independent `div` element could be sandwiched between the parent background element and the "print area" element. This new div could contain a PNG of *just the shirt, without the model.* The `filter` property could then be applied to this stand-alone element, thus leaving everything else in the design space unaffected. I setup a test per this description and it worked like a charm. 

In light of this tests' success, I've abandoned the 2-image pair approach and committed to the `filter` approach. The `filter` system is infinitely more efficient, as each product will only need 3 image-pairs: (a red front/back, a bright green front/back, and a heather-pattern front/back). Last night, I worked on figuring out the `filter` combinations to create each of the 91 colors offered in Gildan. An additional key-value pair will be added to each color in the productColors array, with the value being the necessary `filter` settings for that color. Best of all, this system can be applied to ANY PRODUCT, regardless of style or brand, so long as the base image colors match one of the three (red, safety green, or heather). 

There is one minor issue this system causes that will require some changes in the JSON:

**IMAGE SELECT** - For the new color-select system to work, it was necessary to use three different images: one that is red (most other colors were produced from this image), a bright-safety green (this was needed to create the few bright, vibrant colors such as Lime, Safety Orange, etc), and a heather-pattern image (used for any color that has a heather pattern). Thus, certain color selections will require the browser to swap out base images to ensure the selected color displays correctly. Since each product will have their own unique 3-image pairs, this file information will need to be included in their respective JSON entries. 

The default garment will be either a Gildan G500 or a G200 (I'm not sure yet). In either case, this specific product's info can be included in the JS file for local reference; this way, the client would only need to download the JSON data if they intend to change the product. In those cases, JS would then collect data from the JSON, rather than the local object.

**Comment by NullOperator404** [comment]
UPDATE 2/4/2023

As mentioned in my previous entry, each product will require 3 image pairs (assuming the product is offered in bright vibrant colors as well as heather patterns). After further thought last night, I realized a major fault in this system - the color-select palette will need to identify whether the background image needs to be changed when the user clicks a particular color. For example, if the user clicks "Sport Grey," JS will need to swap the background image to "image3," which is the heather-pattern version before applying the filter. 

This quickly became a bigger and bigger problem the more I thought about it, as it would mean the color-select palette will constantly be having to check the database, since that's where the image data is stored for each product. Furthermore, many of these checks would be completely redundant; if the user selects "red" as the color, then "image1" would be loaded. However, selecting "Royal" means the browser would just re-load "image1" for no reason; Royal is also produced from "image1," and since it's already loaded into the DOM when the user clicked "red," there's no need for the image to be loaded again. 

This is a serious flaw in this approach, as it wholly defeats the point of the `filter` system; the browser would be loading a new image each time *anyway,* so there's no difference between the two approaches; either way, it's constantly having to connect to the server to load an image. 

Earlier tonight I pondered this issue and I may have come up with a work-around. What's needed is some sort of "status" that JS can check each time a color is selected so it knows which image is currently in the DOM. If the new color selected uses the image already loaded, then it should just apply the new `filter` and move on. It would only load a different image if the "status" value doesn't match the value assigned to the color. 

So here's a rough breakdown of how this system would work:

- Each color element in the productColors array will have a key-value pair called  `imageSet.` The value of this key will be either 1, 2, or 3. Each number corresponds to one of the image pairs. 
- When the user selects a color, a local variable called `loadedImageSet` will be assigned the value from the color's `imageSet` key. So selecting "red' will set this var to "1." 
- As part of the color select code block, JS will perform a check (if-else statement) by reading the value of the `imageSet` key in the selected color and comparing it to the value of `loadedImageSet.` If there is a match, JS will simply adjust the `filter` without reloading the image. If there *isn't* a match, JS will load the appropriate image based on the value it read in `imageSet.` After which, the value of `loadedImageSet` would then be updated to reflect the new image pairs currently loaded. 

This should prevent the browser from making unnecessary requests from the server to reload images already loaded in the DOM.

**Comment by NullOperator404** [comment]
UPDATE 2/7/2023

The Change Product functionality is now set up. The system works as intended (though further testing is needed to ensure any bugs are weeded out). Clicking a "Select Product" button inside each product card updates the name of the product as well as updates the image(s). The function also assigns the current product's data object to a global var in case information is needed elsewhere on the page. 

This wraps up most of the functionality for the design space. Most of the work now involves just building out the database and image libraries; the framework is now in place so JS will automatically take into account any additions I make to the product list. 

COLOR PROBLEM

After setting up the "Change Product" function and running a successful test, it's become apparent that the `filter` approach is not going to work as I'd hoped. While it *does* change the colors, the exact hue/shade of the on-screen product does not match it's real-world counterpart, and this is due to other products' base colors not being **exactly** the same hue, shade, contrast, etc as the set used for setting up the system. 

Basically, the base color for a G500 is red, and CSS creates other colors by adjusting the hue, saturation, brightness, etc of that image. However, the 'red' for a G800 or G200 may not be **exactly** the same as the G500 image, and so this will affect the results of the `filter.`

If this system were to be used, it would require extensive work in Photoshop fine-tuning the images so that the `filter` results are accurate. That said, this would probably end up taking as much time as just acquiring individual color images for each product. 

So I've decided to once again abandon the `filter` approach and build a system that uses image pairs. Rather than store image names in the data objects, a more efficient approach I think would be to utilize existing data key-values to build filenames. For example, the front image of a red G500 might be called, "G500_Red_front.png." The product number could be pulled from the object var, and the color name would be pulled from the swatch's ID, as it is now. Concatenation would then be:

`currentProd.ProductNumber + "_" + color + "_front.png`

A duplicate line of code would be used to handle updating the back view:

`currentProd.ProductNumber + "_" + color + "_back.png'

Each product color will have image pairs (a front and  a back). Unfortunately, the downside is that this will mean a labor-intensive, daunting task of collecting all image pairs for every color, for every product.

**Comment by NullOperator404** [comment]
UPDATE 2/11/2023:

The new image pair functionality is in place and working smoothly. Selecting a product updates the color palette to reflect colors only available in the chosen product. 

All that's needed now is to build out the image pairs, which will take some time. This is more of what I consider "busy work," much like building out the clipart libraries. It can be finished once the coding is finished since the framework is now in place. 

With this, the Garment Color/Change Product functionality is finished; the system will work provided there is an entry in the JSON file for the desired product and an image pair to load in the DOM. I am closing this thread, as there's nothing else to be done. 

I've also finished the "Design Tips" feature, which I knew would be brain-dead simple; it's just an array of various tips for users and JS randomly selects one to display in the toolbox when the page is initially loaded. 

That concludes virtually all of the coding & design work for the lab itself other than loose ends that I will address in the Beta phase. 

I am now moving on to the next "step" in the interactive process, where users can click a button to open a pop-up panel and enter quantities, additional products, and get a total price. (See new thread).
